// kuber/build/kuber.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = Kuber;

var _child_process = require("child_process");

function Kuber(client, defaultNamespace = 'default') {
  return {
    client,
    defaultNamespace,

    createNamespace(body) {
      return client.api.v1.namespaces.post({
        body
      });
    },

    createSecret(body, namespace = defaultNamespace) {
      return client.api.v1.namespaces(namespace).secrets.post({
        body
      });
    },

    createDeployment(body, namespace = defaultNamespace) {
      const [apiName, apiVersion] = body.apiVersion.split("/");
      return client.apis[apiName][apiVersion].namespaces(namespace).deployments.post({
        body
      });
    },

    createService(body, namespace = defaultNamespace) {
      return client.api.v1.namespaces(namespace).services.post({
        body
      });
    },

    createIngress(body, namespace = defaultNamespace) {
      return client.apis.extensions.v1beta1.namespaces(namespace).ingresses.post({
        body
      });
    },

    createPersistentVolume(body) {
      return client.api.v1.persistentvolumes.post({
        body
      });
    },

    createPersistentVolumeClaim(body, namespace = defaultNamespace) {
      return client.api.v1.namespaces(namespace).persistentvolumeclaims.post({
        body
      });
    },

    getPods(namespace = defaultNamespace) {
      return client.api.v1.namespaces(namespace).pods.get();
    },

    async getPod(namespace = defaultNamespace, containerName) {
      const {
        body: {
          items: pods
        }
      } = await client.api.v1.namespaces(namespace).pods.get();
      const wordpressPod = pods.find(pod => !!pod.spec.containers.find(container => container.name === containerName));

      if (!wordpressPod) {
        throw new Error(`Failed to find a wordpress pod for site: ${namespace}`);
      }

      return wordpressPod;
    },

    kubectl(cmd) {
      return (0, _child_process.exec)(`kubectl ${cmd}`);
    },

    copy(from, to, options) {
      let cmd = `kubectl cp ${from} ${to}`;

      if (options && options.namespace) {
        cmd += ` --namespace ${options.namespace}`;
      }

      if (options && options.container) {
        cmd += ` --container ${options.container}`;
      }

      return (0, _child_process.exec)(cmd);
    },

    exec(cmd) {
      return (0, _child_process.exec)(`kubectl exec ${cmd}`);
    }

  };
}
