// kuber/build/clients/google.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getGoogleClient = getGoogleClient;
exports.GoogleKuber = GoogleKuber;

var _googleAuthLibrary = require("google-auth-library");

var _kubernetesClient = require("kubernetes-client");

var _kuber = _interopRequireDefault(require("../kuber"));

async function getGoogleClient(zone, clusterId) {
  const authClient = await _googleAuthLibrary.auth.getClient({
    scopes: 'https://www.googleapis.com/auth/cloud-platform'
  });
  const projectId = await _googleAuthLibrary.auth.getProjectId();
  const clusterPath = `projects/${projectId}/locations/${zone}/clusters/${clusterId}`;
  const {
    data: cluster
  } = await authClient.request({
    url: `https://container.googleapis.com/v1beta1/${clusterPath}`
  });

  if (cluster.error) {
    throw new Error(cluster.error.message);
  }

  const {
    token
  } = await authClient.getAccessToken();
  const kubeConfig = {
    apiVersion: "v1",
    kind: "Config",
    preferences: {},
    "current-context": "a",
    contexts: [{
      name: "a",
      context: {
        cluster: "a",
        user: "a"
      }
    }],
    clusters: [{
      name: "a",
      cluster: {
        "certificate-authority-data": cluster.masterAuth.clusterCaCertificate,
        server: `https://${cluster.endpoint}`
      }
    }],
    users: [{
      name: "a",
      user: {
        "auth-provider": {
          config: {
            "access-token": token,
            "cmd-args": "config config-helper --format=json",
            "cmd-path": "/google/google-cloud-sdk/bin/gcloud",
            "token-key": "{.credential.access_token}"
          },
          name: "gcp"
        }
      }
    }]
  };
  const clientApi = new _kubernetesClient.Client1_13({
    config: _kubernetesClient.config.fromKubeconfig(kubeConfig)
  });
  return clientApi.loadSpec();
}

async function GoogleKuber(zone, clusterId, defaultNamespace = 'default') {
  const client = await getGoogleClient(zone, clusterId);
  return (0, _kuber.default)(client, defaultNamespace);
}
